### DISCUSSION TITLE
Allow JSVM Golang templates to output JSON.

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#3973] [Q&A] [answered]
Currently JSVM template rendering isn't capable of outputing JSON directly into the templates.  However I see that this is something that Golang templates can normally do.

```javascript
routerAdd("get", "/", (c) => {  
  const html = require(`${__hooks}/template.js`)(
    `layouts/layout`,
    `home`, 
    `partials/webcam`,
  ).render({
    'webcams': [
      {src: '/webcam', color: 'orange'},
      {color: 'purple'},
    ]
  })
  return c.html(200, html)
})
```

```text
{{$webcams := or (index . "webcams") nil}}
{{.$webcams}}
```

Template output: 

```
 [map[color:orange src:/webcam] map[color:purple]]
```

**Comment by ganigeorgiev** [comment]
There is no `json.Marshal`/`JSON.stringify`-like template function out of the box, if that is what you are asking.

I guess it won't hurt to have it defined by default. Could you elaborate on your use case?

**Comment by dschissler** [comment]
This seems to be doing alright marshaling a map: https://go.dev/play/p/4DVOACQ873o


I'm using Golang template partials that are Alpine.js widgets.  I'd like to have the option to loop each widget with Golang templates (which duplicates the HTML output), or to loop them with Alpine.  I'd like to not be married to any format as the data is winding through my super awesome system below.

I'm having fun researching the current state of web standards.  It's not a bad way to pick up arcane knowledge.

```html
<div class="d-flex flex-row justify-content-center gap-3">
{{- template "webcams" (index . "webcams") -}}
</div>
```

```html
{{define "webcams"}}
  {{template "webcam_includes" }}

  {{- range $i, $webcam := . }}
    {{- template "webcam" $webcam -}}
  {{end}}

{{end}}

{{- /*
  *** Webcam Widget ***
*/ -}}
{{define "webcam"}}
{{$src := "/webcam"}}
{{$size := "small"}}
{{$color := or (index . "color") "green"}}

{{- /* */ -}}
<div webcam
  x-data="{
    id: $id('webcam'),
    src: '{{js $src}}',
    size: '{{js $size}}',
    color: '{{js $color}}',
  }"
  :id="$id('webcam')"
  :style="{
    '--color': color,
  }"

  @webcam:click.window="
  const {id} = $event.detail
  if (id == $root.id) {
    if (size != 'large') {
      $data.size = 'normal'
    }
  } else {
    if (size != 'normal') {
      $data.size = 'small'
    }
  }
  "

  @webcam:dblclick.window="
  const {id} = $event.detail
  if (id == $root.id) {
    $data.size = 'large'
  } else {
    $data.size = 'small'
  }
  ">

  <img
    :src="src"
    :stream="size"
    @click="
      $dispatch('webcam:click', {
        id: $root.id,
      })
    "
    @dblclick="
      $dispatch('webcam:dblclick', {
        id: $root.id,
      })
    ">
</div>
{{end}}

{{define "webcam_includes"}}
<style>
  [webcam] {
    display: inline-block;
  
    > [stream] {
      background-color: var(--color);
      width: var(--width);
      height: var(--height);
    }
  
    > [stream=""], [stream="small"] {
      --width: 192px;
      --height: 144px;
    }
    > [stream="normal"] {
      --width: 320px;
      --height: 240px;
    }
    > [stream="large"] {
      --width: 640px;
      --height: 480px;
    }
  }
</style>
{{end}}
```

**Reply by ganigeorgiev** [reply, marked as answer]
It should work in PocketBase too (under the hood the same `html/template` package is used) if you wrap it in a `script` tag.

**Reply by dschissler** [reply]
```html
<script>
const webcamsJson={{index . "webcams"}}
</script>
```

Thanks!  I didn't realize that it was doing that magic.  That also complicates things a bit since it is a variable name, but what I'm doing is all rather hacky anyways.

I think that we could really use a `marshall` and `unmarshall` template function.  I tried adding it but it's expecting a string, not an object.  Maybe that is why it didn't work.

```javascript
function addFunctions(registry) {
  registry.addFuncs({
    json: function(v) {
      console.log('json')
      return JSON.stringify(v)
    }
  })
}
```

**Reply by ganigeorgiev** [reply]
I actually haven't tested the `$template.addFuncs` as part of a goja script and it may not work because I think it expects `templates.FuncMap`/`map[string]any` and goja may not be able to create a go wrapper func with reflection based only on the js function declaration.

In any case, I'll consider defining globally a `json` function with the next release. A workaround to define custom functions for your templates could be also to create a dummy `Utils` struct and attach methods to it similar to https://github.com/pocketbase/pocketbase/discussions/3431.

**Reply by dschissler** [reply]
I think that it could be useful to be able to get a random 6 digit code as well via an golange template assignment.  This would give funky people an extra option for versioning script injected variables for when it's being duplicated across many template partials.

Although adding JSON straight into the page is easier, it might be necessary to do some manipulations of it for inclusion on the page.  These new waves of tools like HTMX, Alpine.js, etc are pretty cool, but you really have to do some ugly terrible things if you don't have the data in the exact format that you want it.  This would provide an extra tool for that.

**Reply by dschissler** [reply]
@ganigeorgiev An idea for a useful template helper that I saw somewhere in my searches.

`{{obj "key1" "val1" "key2" "val2"}}` ---> `{"key1": "val1", "key2: "val2"}`

`{{arr (obj "key1" "val1" "key2" "val2") (obj "key3" "val3" "key4" "val4") }}` -> etc
